File: surf-api-bukkit/surf-api-bukkit-api/src/main/java/dev/slne/surf/surfapi/bukkit/api/visualizer/visualizer/SurfPatternedVisualizer.java
package dev.slne.surf.surfapi.bukkit.api.visualizer.visualizer;

import dev.slne.surf.surfapi.core.api.packet.entity.entities.display.PacketBlockDisplay;
import java.util.function.Consumer;
import org.bukkit.Location;
import org.bukkit.Material;
import org.jetbrains.annotations.ApiStatus;

public interface SurfPatternedVisualizer extends SurfVisualizer {

  @ApiStatus.Internal
  @ApiStatus.Obsolete
  @Override
  void addVisualLocation(Location visualLocation);

  @ApiStatus.Internal
  @ApiStatus.Obsolete
  @Override
  void addVisualLocation(Location visualLocation, Material material);

  void addVisualPoint(Location point);

  void removeVisualPoint(Location point);

  void setVisualMaterial(Material material);

  void setVisualMaterial(Material material, Consumer<PacketBlockDisplay> consumer);

  void setVisualHeight(int height);

  void setRenderAtHighestPoint(boolean renderAtHighestPoint);
}

File: surf-api-bukkit/surf-api-bukkit-api/src/main/kotlin/dev/slne/surf/surfapi/bukkit/api/visualizer/visualizer/SurfPatternedVisualizer.kt
package dev.slne.surf.surfapi.bukkit.api.visualizer.visualizer

import dev.slne.surf.surfapi.bukkit.api.nms.bridges.packets.entity.BlockDisplaySettings
import org.bukkit.Location
import org.bukkit.Material
import org.jetbrains.annotations.ApiStatus

interface SurfPatternedVisualizer : SurfVisualizer {

    @ApiStatus.Obsolete
    override fun addVisualLocation(
        visualLocation: Location,
        material: Material,
        consumer: BlockDisplaySettings
    )

    fun addVisualPoint(point: Location)

    fun removeVisualPoint(point: Location)

    fun setVisualMaterial(
        material: Material,
        consumer: BlockDisplaySettings.() -> Unit = {}
    )

    fun setVisualHeight(height: Int)

    fun setRenderAtHighestPoint(renderAtHighestPoint: Boolean)
}
